src/compas/datastructures/tree/tree.py

```diff
@@ -0,0 +1,373 @@
+        A generator of the descendants of the tree node.
+
+    """
+
+    DATASCHEMA = {
+        "type": "object",
+        "$recursiveAnchor": True,
+        "properties": {
+            "name": {"type": "string"},
+            "attributes": {"type": "object"},
+            "children": {"type": "array", "items": {"$recursiveRef": "#"}},
+        },
+        "required": ["name", "attributes", "children"],
+    }
+
+    def __init__(self, name=None, attributes=None):
+        super(TreeNode, self).__init__(name=name)
+        self.attributes = attributes or {}
+        self._parent = None
+        self._children = set()
+        self._tree = None
+
+    def __repr__(self):
+        return "<TreeNode {}>".format(self.name)
+
+    @property
+    def is_root(self):
+        return self._parent is None
+
+    @property
+    def is_leaf(self):
+        return not self._children
+
+    @property
+    def is_branch(self):
+        return not self.is_root and not self.is_leaf
+
+    @property
+    def parent(self):
+        return self._parent
+
+    @property
+    def children(self):
+        return self._children
+
+    @property
+    def tree(self):
+        if self.is_root:
+            return self._tree
+        else:
+            return self.parent.tree
+
+    @property
+    def data(self):
+        return {
+            "name": self.name,
+            "attributes": self.attributes,
+            "children": [child.data for child in self.children],
+        }
+
+    @classmethod
+    def from_data(cls, data):
+        node = cls(data["name"], data["attributes"])
+        for child in data["children"]:
+            node.add(cls.from_data(child))
+        return node
+
+    def add(self, node):
+        """
+        Add a child node to this node.
+
+        Parameters
+        ----------
+        node : :class:`~compas.datastructures.TreeNode`
+            The node to add.
+
+        Returns
+        -------
+        None
+
+        Raises
+        ------
+        TypeError
+            If the node is not a :class:`~compas.datastructures.TreeNode` object.
+
+        """
+        if not isinstance(node, TreeNode):
+            raise TypeError("The node is not a TreeNode object.")
+        self._children.add(node)
+        node._parent = self
+
+    def remove(self, node):
+        """
+        Remove a child node from this node.
+
+        Parameters
+        ----------
+        node : :class:`~compas.datastructures.TreeNode`
+            The node to remove.
+
+        Returns
+        -------
+        None
+
+        """
+        self._children.remove(node)
+        node._parent = None
+
+    @property
+    def ancestors(self):
+        this = self
+        while this:
+            yield this
+            this = this.parent
+
+    @property
+    def descendants(self):
+        for child in self.children:
+            yield child
+            for descendant in child.descendants:
+                yield descendant
+
```

Comment:
this implies the children are returned in "depth-first" order, no? either way, could you add a comment about this in the attributes section of the class docstring?